src/dbnode/storage/index/block.go

```diff
@@ -543,15 +545,20 @@ func (b *block) queryWithSpan(
 		// Ensure that the block contains any of the relevant time segments for the query range.
 		doc := iter.Current()
 		if md, ok := doc.Metadata(); ok && md.OnIndexSeries != nil {
```

Comment:
I wonder if it'd be worth, for preventing mistaken calls directly to `metadata.OnIndexSeries` in the future, having a separate interface that `Entry` implements called something like `OnIndexSeriesReference` that just has a single func for `GetOnIndexSeries` (i.e. `ReconciledOnIndexSeries`). And then we have `Metadata` have that so a caller is forced to always go through this reconciled API instead of referencing the field directly

Reply:
Do we always want to reconcile though? I'm not sure that's true from looking at other users of `metadata.OnIndexSeries`. For example, when we call `NeedsIndexGarbageCollected` it seems like it specifically doesn't reconcile because it's not worth the extra cost:

```go
// NeedsIndexGarbageCollected checks if the entry is eligible to be garbage collected
// from the index. Otherwise returns false.
func (entry *Entry) NeedsIndexGarbageCollected() bool {
	// This is a cheaper check that loading the entry from the shard again
	// which makes it cheaper to run frequently.
	// It may not be as accurate, but it's fine for an approximation since
	// only a single series in a segment needs to return true to trigger an
	// index segment to be garbage collected.
	if entry.insertTime.Load() == 0 {
		return false // Not inserted, does not need garbage collection.
	}
	// Check that a write is not potentially pending and the series is empty.
	return entry.ReaderWriterCount() == 0 && entry.Series.IsEmpty()
}
```

If we change the interface here, we're always going to reconcile when insertTime is 0.

Reply:
Hm yeah good point that we intentionally don't want to reconcile in other cases.